core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java

```diff
@@ -31,10 +30,8 @@
  * Reads spill files written by {@link UnsafeSorterSpillWriter} (see that class for a description
  * of the file format).
  */
-public final class UnsafeSorterSpillReader extends UnsafeSorterIterator {
-  private static final Logger logger = LoggerFactory.getLogger(UnsafeSorterSpillReader.class);
+public final class UnsafeSorterSpillReader extends UnsafeSorterIterator implements Closeable {
```

Comment:
Of the potential leaks that were discovered, I think that this is the only one to be concerned about: it looks like `UnsafeSorterSpillReader` might leak an open `FileInputStream` if an exception occurred while reading records. The fix implemented here is to add a `close()` method for safely closing the reader's streams.

I chose to move the spill deletion logic out of the reader itself since it appears to be redundant with spill deletion code that lives elsewhere. We should audit the existing code to make sure that the chain of responsibility for cleaning up spill files is clearly defined.